src/dataflow-types/src/explain.rs

```diff
@@ -231,3 +231,39 @@ impl<'a> ViewFormatter<OptimizedMirRelationExpr> for DataflowGraphFormatter<'a>
         fmt::Display::fmt(&explain, f)
     }
 }
+
+pub struct TimestampExplanation<T> {
+    pub timestamp: T,
+    pub has_table: bool,
+    pub get_local_read_ts: Option<T>,
+    pub since: Vec<T>,
+    pub upper: Vec<T>,
+    pub sources: Vec<TimestampSource<T>>,
+}
+
+pub struct TimestampSource<T> {
+    pub name: String,
+    pub read_frontier: Vec<T>,
+    pub write_frontier: Vec<T>,
+}
```

Comment:
Not in charge here, but I'd love to see some doccomments on these to explain them and their fields. I can guess at most of these, but e.g. `has_table` is a bit esoteric, and `get_local_read_ts` probably only makes sense to folks who knows the method names used to acquire timestamps.

Reply:
Done.